File: packages/uns-cli/templates/cron/src/examples/cron-example.ts
/**
 * Change this file according to your specifications and rename it to index.ts
 */
import { UnsProxyProcess, ConfigFile, logger, UnsEvents } from "@uns-kit/core";
import { registerAttributeDescriptions, registerObjectTypeDescriptions } from "@uns-kit/core/uns/uns-dictionary-registry.js";
import "@uns-kit/cron";
import { type UnsProxyProcessWithCron } from "@uns-kit/cron";
import { UnsTopics } from "@uns-kit/core/uns/uns-topics.js";
import { UnsPacket } from "@uns-kit/core/uns/uns-packet.js";
import {
  GeneratedObjectTypes,
  GeneratedAttributes,
  GeneratedAttributesByType,
  GeneratedAttributeDescriptions,
  GeneratedObjectTypeDescriptions,
} from "./uns/uns-dictionary.generated.js";
import { GeneratedPhysicalMeasurements } from "./uns/uns-measurements.generated.js";


/**
 * Load the configuration from a file.
 * On the server, this file is provided by the `uns-datahub-controller`.
 * In the development environment, you are responsible for creating and maintaining this file and its contents.
 */
const config = await ConfigFile.loadConfig();
registerObjectTypeDescriptions(GeneratedObjectTypeDescriptions);
registerAttributeDescriptions(GeneratedAttributeDescriptions);

/**
 * Connect to the output broker and create a crontab proxy
 */
const unsProxyProcess = new UnsProxyProcess(config.infra.host!, {processName:config.uns.processName}) as UnsProxyProcessWithCron;;
const mqttOutput = await unsProxyProcess.createUnsMqttProxy((config.output?.host)!, "templateUnsRttOutput", config.uns.instanceMode!, config.uns.handover!, { publishThrottlingDelay: 1000});
const cronInput = await unsProxyProcess.createCrontabProxy("* * * * * *");

/**
 * Event listener for cron events.
 * On each cron event, publish a message to the MQTT output broker.
 */
cronInput.event.on("cronEvent", async (event: UnsEvents["cronEvent"]) => {
  try {
    const time = UnsPacket.formatToISO8601(new Date());
    const numberValue: number = 42;
    const topic: UnsTopics = "example/";
    const asset = "asset";
    const assetDescription = "Sample asset";

    const dataGroup = "sensor";

    mqttOutput.publishMqttMessage({
      topic,
      asset,
      assetDescription,
      objectType: GeneratedObjectTypes.equipment,
      objectId: "main",
      attributes: [
        {
          attribute: GeneratedAttributesByType["energy-resource"]["current"],
          data: { dataGroup, time, value: numberValue, uom: GeneratedPhysicalMeasurements.Ampere }
        }
      ],
    });
  } catch (error) {
    const reason = error instanceof Error ? error : new Error(String(error));
    logger.error(`Error publishing message to MQTT: ${reason.message}`);
    throw error;
  }
});

File: packages/uns-py/src/uns_kit/templates/cron/src/cron_example.py
import asyncio
import logging
from datetime import datetime, timezone
from pathlib import Path

from uns_kit.core.config_file import ConfigFile
from uns_kit.core.logger import configure_logger
from uns_kit.core.proxy_process import UnsProxyProcess


configure_logger(settings={"level": "INFO", "console": True})
logger = logging.getLogger("uns_kit").getChild(__name__)


def load_config() -> dict:
    cfg_path = Path("config.json")
    if cfg_path.exists():
        return ConfigFile.load_config(cfg_path)
    return {"infra": {"host": "localhost"}, "uns": {"processName": "uns-process"}}


async def run() -> None:
    config = load_config()
    infra = config.get("infra") or {}
    uns = config.get("uns") or {}

    process = UnsProxyProcess(
        infra.get("host") or "localhost",
        {"processName": uns.get("processName") or "uns-cron"},
    )
    await process.start()
    mqtt_output = await process.create_mqtt_proxy("templateUnsRttOutput")
    cron = await process.create_cron_proxy("* * * * * *")

    async def handle_cron(_event):
        try:
            await mqtt_output.publish_mqtt_message(
                {
                    "topic": "example/site/area/line/",
                    "asset": "demo-asset",
                    "objectType": "energy-resource",
                    "objectId": "main-bus",
                    "attributes": [
                        {
                            "attribute": "current",
                            "description": "Cron sample current",
                            "data": {
                                "time": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
                                "value": 42,
                                "uom": "Ampere",
                                "dataGroup": "sensor",
                            },
                        }
                    ],
                }
            )
        except Exception as exc:
            logger.error("Error publishing message to MQTT: %s", exc)
            raise

    cron.event.on("cronEvent", handle_cron)

    try:
        await asyncio.Event().wait()
    finally:
        await mqtt_output.flush()
        await mqtt_output.close()
        await process.stop()


if __name__ == "__main__":
    asyncio.run(run())
